docs/cpp/source/notes/tensor_cuda_stream.rst

```diff
@@ -0,0 +1,173 @@
+
+.. code-block:: cpp
+  
+  CUDAStream getDefaultCUDAStream(DeviceIndex device_index = -1);
+
+.. tip::
+  
+  The default stream is where most computation occurs when you aren't explicitly using streams.
+
+3. Get the current CUDA stream, for the CUDA device with index ``device_index``, or for the current device if no device index is passed. 
+
+.. code-block:: cpp
+  
+  CUDAStream getCurrentCUDAStream(DeviceIndex device_index = -1);
+
+.. tip::
+  
+  The current CUDA stream will usually be the default CUDA stream for the device, but it may be different if someone
+  called ``setCurrentCUDAStream`` or used ``StreamGuard`` or ``CUDAStreamGuard``.
+
+
+
+Set CUDA stream
+***************
+
+Pytorch C++ API provides the following ways to set CUDA stream:
+
+1. Set the current stream on the device of the passed in stream to be the passed in stream.
+
+.. code-block:: cpp
+
+  void setCurrentCUDAStream(CUDAStream stream);
+
+.. attention::
+
+  This function may have nosthing to do with the current device. It only changes the current stream on the stream's device.
+  We recommend using ``CUDAStreamGuard``, instead, since it switches to the stream's device and makes it the current stream on that device.
+  ``CUDAStreamGuard`` will also restore the current device and stream when it's destroyed
+
+2. Use ``CUDAStreamGuard`` to switch to a CUDA stream within a scope, it is defined in `CUDAStreamGuard.h`_
+
+.. tip::
+
+  Use ``CUDAMultiStreamGuard`` if you need to set streams on multiple CUDA devices.
+
+CUDA Stream Usage Examples
+**************************
+
+1. Acquiring and setting a CUDA stream
+
+.. code-block:: cpp
+
+  // get a new stream from CUDA stream pool on current device (device 0)
+  // set current stream with this new stream and get current stream
+  torch::cuda::CUDAStream myStream = at::cuda::getStreamFromPool();
+  torch::cuda::setCurrentCUDAStream(myStream);
+  torch::cuda::CUDAStream curStream = torch::cuda::getCurrentCUDAStream();
+  assert(curStream == myStream);
+
+  // get the default CUDA stream on current device (device 0)
+  // set current stream with default stream and get current stream
+  torch::cuda::CUDAStream defaultStream = torch::cuda::getDefaultCUDAStream();
+  torch::cuda::setCurrentCUDAStream(defaultStream);
+  curStream = torch::cuda::getCurrentCUDAStream();
+  assert(curStream == defaultStream);
+  assert(myStream != defaultStream)
+
+.. attention::
+  
+  Above code is running on the same device/gpu (say device 0). `setCurrentCUDAStream` works as expected and set current CUDA stream correctly
+  on device 0, this is because all the streams are on the same device. However, say above code is running on device 0 and we get `myStream`
+  from device 1, then `torch::cuda::setCurrentCUDAStream(myStream)` will set `myStream` as current stream on device 1, not device 0.
+
+
+2. Use various CUDA Guard to set CUDA stream
+
+.. code-block:: cpp
+
+   // create a vector of CUDA stream from current device (device 0)
```

Comment:
I'm not really sure I get what this example is supposed to show?

Reply:
This example is to list different usage of guard.  It listed the changes of streams and devices in each step to demonstrate how the guard works.